File: src/java/org/apache/fop/events/EventFormatter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.events;

import java.util.Locale;
import java.util.Map;
import java.util.MissingResourceException;
import java.util.ResourceBundle;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.apache.fop.util.XMLResourceBundle;
import org.apache.fop.util.text.AdvancedMessageFormat;
import org.apache.fop.util.text.AdvancedMessageFormat.Part;
import org.apache.fop.util.text.AdvancedMessageFormat.PartFactory;

/**
 * Converts events into human-readable, localized messages.
 */
public final class EventFormatter {

    private static final Pattern INCLUDES_PATTERN = Pattern.compile("\\{\\{.+\\}\\}");
    
    private static ResourceBundle defaultBundle = XMLResourceBundle.getXMLBundle(
            EventFormatter.class.getName(), EventFormatter.class.getClassLoader());
    
    private static Log log = LogFactory.getLog(EventFormatter.class);
    
    private EventFormatter() {
        //utility class
    }
    
    /**
     * Formats an event using the default locale.
     * @param event the event
     * @return the formatted message
     */
    public static String format(Event event) {
        ResourceBundle bundle = null;
        String groupID = event.getEventGroupID();
        if (groupID != null) {
            try {
                 bundle = XMLResourceBundle.getXMLBundle(
                        groupID,
                        EventFormatter.class.getClassLoader());
            } catch (MissingResourceException mre) {
                if (log.isTraceEnabled()) {
                    log.trace("No XMLResourceBundle for " + groupID + " available.");
                }
            }
        }
        if (bundle == null) {
            bundle = defaultBundle;
        }
        return format(event, bundle);
    }
    
    /**
     * Formats an event using a given locale.
     * @param event the event
     * @param locale the locale
     * @return the formatted message
     */
    public static String format(Event event, Locale locale) {
        ResourceBundle bundle = null;
        String groupID = event.getEventGroupID();
        if (groupID != null) {
            try {
                 bundle = XMLResourceBundle.getXMLBundle(
                        groupID, locale,
                        EventFormatter.class.getClassLoader());
            } catch (MissingResourceException mre) {
                if (log.isTraceEnabled()) {
                    log.trace("No XMLResourceBundle for " + groupID + " available.");
                }
            }
        }
        if (bundle == null) {
            bundle = XMLResourceBundle.getXMLBundle(
                    EventFormatter.class.getName(),
                    locale,
                    EventFormatter.class.getClassLoader());
        }
        return format(event, bundle);
    }

    private static String format(Event event, ResourceBundle bundle) {
        String template = bundle.getString(event.getEventID());
        return format(event, processIncludes(template, bundle));
    }

    private static String processIncludes(String template, ResourceBundle bundle) {
        CharSequence input = template;
        int replacements;
        StringBuffer sb;
        do {
            sb = new StringBuffer(Math.max(16, input.length()));
            replacements = processIncludesInner(input, sb, bundle);
            input = sb;
        } while (replacements > 0);
        String s = sb.toString();
        return s;
    }

    private static int processIncludesInner(CharSequence template, StringBuffer sb,
            ResourceBundle bundle) {
        int replacements = 0;
        Matcher m = INCLUDES_PATTERN.matcher(template);
        while (m.find()) {
            String include = m.group();
            include = include.substring(2, include.length() - 2);
            m.appendReplacement(sb, bundle.getString(include));
            replacements++;
        }
        m.appendTail(sb);
        return replacements;
    }

    /**
     * Formats the event using a given pattern. The pattern needs to be compatible with
     * {@link AdvancedMessageFormat}.
     * @param event the event
     * @param pattern the pattern (compatible with {@link AdvancedMessageFormat})
     * @return the formatted message
     */
    public static String format(Event event, String pattern) {
        AdvancedMessageFormat format = new AdvancedMessageFormat(pattern);
        Map params = new java.util.HashMap(event.getParams());
        params.put("source", event.getSource());
        params.put("severity", event.getSeverity());
        return format.format(params);
    }
    
    private static class LookupFieldPart implements Part {
        
        private String fieldName;
        
        public LookupFieldPart(String fieldName) {
            this.fieldName = fieldName;
        }

        public boolean isGenerated(Map params) {
            return getKey(params) != null;
        }

        public void write(StringBuffer sb, Map params) {
            sb.append(defaultBundle.getString(getKey(params)));
        }

        private String getKey(Map params) {
            return (String)params.get(fieldName);
        }
        
        /** {@inheritDoc} */
        public String toString() {
            return "{" + this.fieldName + ", lookup}";
        }
        
    }
    
    /** PartFactory for lookups. */
    public static class LookupFieldPartFactory implements PartFactory {

        /** {@inheritDoc} */
        public Part newPart(String fieldName, String values) {
            return new LookupFieldPart(fieldName);
        }

        /** {@inheritDoc} */
        public String getFormat() {
            return "lookup";
        }
        
    }

}

File: dotnet/src/Fop.Events/EventFormatter.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using Fop.Util.Text;

namespace Fop.Events;

/// <summary>
/// Converts events into human-readable messages.
/// <para>
/// Port of <c>org.apache.fop.events.EventFormatter</c>. The Java version resolved message templates
/// from XML resource bundles. Those bundle files are not part of this port, so template lookup is
/// delegated to a pluggable <see cref="IEventModelMessageSource"/> (see <see cref="MessageSource"/>).
/// The template is then rendered with <see cref="AdvancedMessageFormat"/>, exactly as in Java.
/// </para>
/// </summary>
public static partial class EventFormatter
{
#if NET7_0_OR_GREATER
    [GeneratedRegex("\\{\\{.+\\}\\}")]
    private static partial Regex IncludesPatternGenerator();

    private static Regex IncludesPattern => IncludesPatternGenerator();
#else
    private static readonly Regex IncludesPattern = new("\\{\\{.+\\}\\}");
#endif

    static EventFormatter()
    {
        // Replaces the Java service-provider registration of the "lookup" PartFactory.
        AdvancedMessageFormat.RegisterPartFactory(new LookupFieldPartFactory());
    }

    /// <summary>
    /// The source of message templates. Defaults to an empty
    /// <see cref="InMemoryEventModelMessageSource"/>; assign a populated source (or a custom
    /// implementation) before formatting events.
    /// </summary>
    public static IEventModelMessageSource MessageSource { get; set; } =
        new InMemoryEventModelMessageSource();

    /// <summary>Formats an event using the event's own culture.</summary>
    /// <param name="event">The event.</param>
    /// <returns>The formatted message.</returns>
    public static string Format(Event @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        return Format(@event, @event.Culture);
    }

    /// <summary>Formats an event using a given culture.</summary>
    /// <param name="event">The event.</param>
    /// <param name="culture">The culture (may be <c>null</c>).</param>
    /// <returns>The formatted message.</returns>
    public static string Format(Event @event, CultureInfo? culture)
    {
        ArgumentNullException.ThrowIfNull(@event);
        string key = @event.EventKey;
        string? template = MessageSource.GetTemplate(@event.EventGroupId, key, culture);
        if (template == null)
        {
            // Java rendered "Missing bundle..." when no bundle existed for the group; a missing key
            // within an existing bundle threw. Both observable outcomes collapse into this single,
            // clear fallback message here.
            template = "Missing bundle. Can't lookup event key: '" + key + "'.";
            return Format(@event, template);
        }
        return Format(@event, ProcessIncludes(template, @event.EventGroupId, culture));
    }

    private static string ProcessIncludes(string template, string? groupId, CultureInfo? culture)
    {
        string input = template;
        int replacements;
        StringBuilder sb;
        do
        {
            sb = new StringBuilder(Math.Max(16, input.Length));
            replacements = ProcessIncludesInner(input, sb, groupId, culture);
            input = sb.ToString();
        }
        while (replacements > 0);
        return sb.ToString();
    }

    private static int ProcessIncludesInner(string template, StringBuilder sb, string? groupId,
        CultureInfo? culture)
    {
        int replacements = 0;
        int last = 0;
        foreach (Match m in IncludesPattern.Matches(template))
        {
            string include = m.Value;
            include = include.Substring(2, include.Length - 4);
            string? replacement = MessageSource.GetTemplate(groupId, include, culture);
            sb.Append(template, last, m.Index - last);
            sb.Append(replacement ?? string.Empty);
            last = m.Index + m.Length;
            replacements++;
        }
        sb.Append(template, last, template.Length - last);
        return replacements;
    }

    /// <summary>
    /// Formats the event using a given pattern. The pattern needs to be compatible with
    /// <see cref="AdvancedMessageFormat"/>.
    /// </summary>
    /// <param name="event">The event.</param>
    /// <param name="pattern">The pattern (compatible with <see cref="AdvancedMessageFormat"/>).</param>
    /// <returns>The formatted message.</returns>
    public static string Format(Event @event, string pattern)
    {
        ArgumentNullException.ThrowIfNull(@event);
        ArgumentNullException.ThrowIfNull(pattern);
        var format = new AdvancedMessageFormat(pattern);
        var parameters = new Dictionary<string, object?>(@event.Parameters)
        {
            ["source"] = @event.Source,
            ["severity"] = @event.Severity,
            ["groupID"] = @event.EventGroupId,
            ["locale"] = @event.Culture,
        };
        return format.Format(parameters);
    }
}
